File: internal/definition/submission.go
package definition

// Visible reports, for every field key, whether the field is shown for data
// (spec §5.4). It shares one evaluator with ValidateSubmission so the two can
// never disagree.
func Visible(f Form, data map[string]any) map[string]bool {
	visible, _, _ := evaluate(f, data)
	return visible
}

// ValidateSubmission cleans data (drops unknown keys, hidden fields and
// not-provided values; trims strings; normalizes numbers and lists) and
// validates every visible field. It returns the cleaned data or a
// *ValidationError whose problems use paths "data.<key>" in field order.
func ValidateSubmission(f Form, data map[string]any) (map[string]any, error) {
	_, clean, ps := evaluate(f, data)
	if err := ps.err(); err != nil {
		return nil, err
	}
	return clean, nil
}

// evaluate walks the fields in declaration order. A field's condition is
// evaluated against the cleaned values of earlier fields; a hidden controller
// hides its dependants; an invalid value counts as absent.
func evaluate(f Form, data map[string]any) (map[string]bool, map[string]any, problems) {
	visible := make(map[string]bool, len(f.Fields))
	clean := make(map[string]any, len(f.Fields))
	types := make(map[string]FieldType, len(f.Fields))
	var ps problems
	for _, fld := range f.Fields {
		types[fld.Key] = fld.Type
		vis := true
		if c := fld.ShowIf; c != nil {
			if !visible[c.Field] {
				vis = false
			} else {
				val, present := clean[c.Field]
				vis = holds(*c, types[c.Field], val, present)
			}
		}
		visible[fld.Key] = vis
		if !vis {
			continue
		}
		path := "data." + fld.Key
		val, provided, msg := checkValue(fld.Type, fld.Options, fld.Validation, data[fld.Key])
		switch {
		case msg != "":
			ps.add(path, msg)
		case !provided:
			if fld.Required {
				ps.add(path, "is required")
			}
		case fld.Type == FieldCheckbox && fld.Required && val == false:
			ps.add(path, "must be checked")
		default:
			clean[fld.Key] = val
		}
	}
	return visible, clean, ps
}

// holds evaluates a condition against the controller's cleaned value.
func holds(c Condition, ctrl FieldType, val any, present bool) bool {
	multi := ctrl == FieldMultiselect
	switch {
	case c.Equals != nil:
		if !present {
			return false
		}
		if multi {
			return containsJSON(val.([]any), c.Equals)
		}
		return jsonEqual(val, c.Equals)
	case c.NotEquals != nil:
		if !present {
			return true
		}
		if multi {
			return !containsJSON(val.([]any), c.NotEquals)
		}
		return !jsonEqual(val, c.NotEquals)
	case c.In != nil:
		if !present {
			return false
		}
		if multi {
			for _, item := range val.([]any) {
				if containsJSON(c.In, item) {
					return true
				}
			}
			return false
		}
		return containsJSON(c.In, val)
	}
	return true
}

func containsJSON(list []any, v any) bool {
	for _, it := range list {
		if jsonEqual(it, v) {
			return true
		}
	}
	return false
}

File: src/openforms/definition/submission.py
"""Visibility and submission validation (spec §5.4). One evaluator serves both,
so ``visible`` and ``validate_submission`` can never disagree."""

from __future__ import annotations

from typing import Any

from .problems import Problems
from .types import FIELD_CHECKBOX, FIELD_MULTISELECT, Condition, Form
from .values import check_value, json_equal


def visible(f: Form, data: dict[str, Any] | None) -> dict[str, bool]:
    """Whether each field is shown for ``data`` (every field key is present)."""
    vis, _, _ = _evaluate(f, data or {})
    return vis


def validate_submission(f: Form, data: dict[str, Any] | None) -> dict[str, Any]:
    """Return cleaned data (unknown keys, hidden fields and blanks dropped; strings
    trimmed) or raise ValidationError with ``data.<key>`` paths in field order."""
    _, clean, ps = _evaluate(f, data or {})
    ps.raise_if_any()
    return clean


def _evaluate(f: Form, data: dict[str, Any]) -> tuple[dict[str, bool], dict[str, Any], Problems]:
    vis: dict[str, bool] = {}
    clean: dict[str, Any] = {}
    types: dict[str, str] = {}
    ps = Problems()
    for fld in f.field_list():
        types[fld.key] = fld.type
        shown = True
        c = fld.show_if
        if c is not None:
            if not vis.get(c.field, False):
                shown = False
            else:
                shown = _holds(c, types.get(c.field, ""), clean.get(c.field), c.field in clean)
        vis[fld.key] = shown
        if not shown:
            continue
        path = "data." + fld.key
        val, provided, msg = check_value(fld.type, fld.options, fld.validation, data.get(fld.key))
        if msg:
            ps.add(path, msg)
        elif not provided:
            if fld.required:
                ps.add(path, "is required")
        elif fld.type == FIELD_CHECKBOX and fld.required and val is False:
            ps.add(path, "must be checked")
        else:
            clean[fld.key] = val
    return vis, clean, ps


def _contains(items: list[Any], v: Any) -> bool:
    return any(json_equal(it, v) for it in items)


def _holds(c: Condition, ctrl: str, val: Any, present: bool) -> bool:
    multi = ctrl == FIELD_MULTISELECT
    if c.equals is not None:
        if not present:
            return False
        return _contains(val, c.equals) if multi else json_equal(val, c.equals)
    if c.not_equals is not None:
        if not present:
            return True
        return not _contains(val, c.not_equals) if multi else not json_equal(val, c.not_equals)
    if c.in_ is not None:
        if not present:
            return False
        if multi:
            return any(_contains(c.in_, item) for item in val)
        return _contains(c.in_, val)
    return True
